### PR TITLE
8296709: WARNING: JNI call made without checking exceptions

### PR DESCRIPTION
**Pull request by RealLucy**
This small change fixes a warning that may pop up during runtime. May I please request reviews? Thank you!

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8296709](https://bugs.openjdk.org/browse/JDK-8296709): WARNING: JNI call made without checking exceptions

### Reviewers
 * [Kevin Walls](https://openjdk.org/census#kevinw) (@kevinjwalls - Committer) ⚠️ Review applies to [18aec934](https://git.openjdk.org/jdk/pull/11083/files/18aec9344cb567d98da4e02f62e2d089569bf503)
 * [Serguei Spitsyn](https://openjdk.org/census#sspitsyn) (@sspitsyn - **Reviewer**) ⚠️ Review applies to [cea8cf87](https://git.openjdk.org/jdk/pull/11083/files/cea8cf879f64b4318d8a1c12f435de2bda77dfca)
 * [Christoph Langer](https://openjdk.org/census#clanger) (@RealCLanger - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11083/head:pull/11083` \
`$ git checkout pull/11083`

Update a local copy of the PR: \
`$ git checkout pull/11083` \
`$ git pull https://git.openjdk.org/jdk pull/11083/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11083`

View PR using the GUI difftool: \
`$ git pr show -t 11083`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11083.diff">https://git.openjdk.org/jdk/pull/11083.diff</a>

</details>

### CODE REVIEW

**Review by kevinjwalls — APPROVED** [review]
Looks good I think.  Got to check for those exceptions!

The example invocation also shows the me "exceeds capacity" warning, which must relate to the above WITH_LOCAL_REFS and call to createLocalRefSpace.  Another time maybe. 8-)

**Comment by RealLucy** [comment]
Thanks for the review, Kevin! 
I'm looking into the "exceeds capacity" warning separately. Need to find out who is holding so many local references, just for a java -version. Once I have enough knowledge, I'll file a separate bug.

**Review by sspitsyn — APPROVED** [review]
Looks good.
Thank you for taking care about this.
Serguei

**Comment by dholmes-ora** [comment]
Can the test be augmented to catch this case too?

**Comment by RealCLanger** [comment]
Maybe you could also call "bin/java -Xcheck:jni -agentlib:jdwp=transport=dt_socket,server=y,suspend=n -version" which would show the error? Or add this as another test for serviceability?

**Comment by dholmes-ora** [comment]
I was suggesting what @RealCLanger suggests in regards to the test - add a new case.

**Comment by RealLucy** [comment]
> I was suggesting what @RealCLanger suggests in regards to the test - add a new case.

Ahh, now I understand. I will add the suggested test modification to this PR.

**Comment by RealLucy** [comment]
I've added a separate test which does the same but with an active jdwp agent. The test will fail until the warning 
`WARNING: JNI local refs: 921, exceeds capacity: 32`
is fixed. I'm not sure how to do this. Just increasing the capacity doesn't seem to be the right thing.

File: test/jdk/tools/launcher/TestXcheckJdwpJNIWarnings.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright (c) 2017, Oracle and/or its affiliates. All rights reserved.
```
Comment (RealCLanger) [inline test/jdk/tools/launcher/TestXcheckJdwpJNIWarnings.java:2]: Add the current year to copyright.

File: test/jdk/tools/launcher/TestXcheckJdwpJNIWarnings.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+ * questions.
+ */
+
+import java.io.File;
+import java.io.IOException;
+
+/**
+ * @test
+ * @bug 8187442
+ * @summary Launching app shouldn't produce any jni warnings.
```
Comment (RealCLanger) [inline test/jdk/tools/launcher/TestXcheckJdwpJNIWarnings.java:30]: Maybe better: "Launching app with debug agent shouldn't produce any jni warnings." ?

**Review by RealCLanger — CHANGES_REQUESTED** [review]
Looks good. I commented on two nits.

If you don't have a quick idea on how to fix `WARNING: JNI local refs: 921, exceeds capacity: 32`, you could open a bug for it (if it isn't) and right away exclude the test with reference to the new bug.
The alternative would be to filter out the local refs warning but then I guess this will be forgotten.
After all, either way is not perfect...

**Comment by dholmes-ora** [comment]
Couldn't you simply add a second `doExec` statement in the existing test rather than copy the entire test file?

File: test/jdk/tools/launcher/TestXcheckJdwpJNIWarnings.java (hunk 2 above)
Comment (RealLucy) [reply test/jdk/tools/launcher/TestXcheckJdwpJNIWarnings.java:30]: Agreed. Text updated accordingly.

File: test/jdk/tools/launcher/TestXcheckJdwpJNIWarnings.java (hunk 1 above)
Comment (RealLucy) [reply test/jdk/tools/launcher/TestXcheckJdwpJNIWarnings.java:2]: File is new. Therefore changed year to 2022.

**Comment by RealLucy** [comment]
> Couldn't you simply add a second `doExec` statement in the existing test rather than copy the entire test file?

Actually, I first did it that way. And I did not like it. It's not just the doExec, you need to duplicate the checks as well. So far, it's just personal preference. 

The test case with active jdwp agent produces two warnings, but I have a fix for only one of the warnings. @RealCLanger suggested to create a bug for the unresolved warning and problem-list the test case until the bug is fixed. If this is the way to go, I need a separate test file.

**Comment by RealCLanger** [comment]
> > Couldn't you simply add a second `doExec` statement in the existing test rather than copy the entire test file?
> 
> Actually, I first did it that way. And I did not like it. It's not just the doExec, you need to duplicate the checks as well. So far, it's just personal preference.
> 
> The test case with active jdwp agent produces two warnings, but I have a fix for only one of the warnings. @RealCLanger suggested to create a bug for the unresolved warning and problem-list the test case until the bug is fixed. If this is the way to go, I need a separate test file.

Hm, I also don't like the code duplication...

You could use 2 @run tags. Via the run tags you can create two runs of the test, one like the current "default" and antoher one passing the debugging agent option as additional parameter which you have to hand to the the java child process then, too. Generally you'd expect the same results, e.g. good return code, the hello world and no warnings.

With that approach you can then exclude the failing run via exclude list.

**Comment by RealLucy** [comment]
> Hm, I also don't like the code duplication...
 
Code duplication is gone. Just one test file with two @run tags.

**Comment by kevinjwalls** [comment]
To see where the JNI local refs warning comes from, I set a breakpoint in JavaThread::print_stack_on as that is called when the warning happens, although it doesn't print anything as there is no Java frame.
...
(gdb) cont
Continuing.
WARNING: JNI local refs: 442, exceeds capacity: 32

Breakpoint 1, 0x00007ffff6efb8c0 in JavaThread::print_stack_on(outputStream*) ()
   from /scratch/kwalls/jdk/jdk-17.0.3/lib/server/libjvm.so
(gdb) bt
#0  0x00007ffff6efb8c0 in JavaThread::print_stack_on(outputStream*) ()
   from /scratch/kwalls/jdk/jdk-17.0.3/lib/server/libjvm.so
#1  0x00007ffff6a0f38c in checked_jni_GetJavaVM () from /scratch/kwalls/jdk/jdk-17.0.3/lib/server/libjvm.so
#2  0x00007ffff7e8dfa8 in transport_startTransport () from /scratch/kwalls/jdk/jdk-17.0.3/lib/libjdwp.so
#3  0x00007ffff7e76f77 in startTransport () from /scratch/kwalls/jdk/jdk-17.0.3/lib/libjdwp.so
#4  0x00007ffff7e75069 in bagEnumerateOver () from /scratch/kwalls/jdk/jdk-17.0.3/lib/libjdwp.so
#5  0x00007ffff7e77948 in initialize () from /scratch/kwalls/jdk/jdk-17.0.3/lib/libjdwp.so
#6  0x00007ffff7e783f0 in cbEarlyVMInit () from /scratch/kwalls/jdk/jdk-17.0.3/lib/libjdwp.so
#7  0x00007ffff6b5e756 in JvmtiExport::post_vm_initialized() ()
   from /scratch/kwalls/jdk/jdk-17.0.3/lib/server/libjvm.so
#8  0x00007ffff6f00f54 in Threads::create_vm(JavaVMInitArgs*, bool*) ()
   from /scratch/kwalls/jdk/jdk-17.0.3/lib/server/libjvm.so
#9  0x00007ffff6a007d1 in JNI_CreateJavaVM () from /scratch/kwalls/jdk/jdk-17.0.3/lib/server/libjvm.so
...

Specifically breaking in JNIHandles::make_local I see a lot of activity like:

Breakpoint 2, 0x00007ffff6a240c4 in JNIHandles::make_local(oopDesc*) ()
   from /scratch/kwalls/jdk/jdk-17.0.3/lib/server/libjvm.so
#0  0x00007ffff6a240c4 in JNIHandles::make_local(oopDesc*) () from /scratch/kwalls/jdk/jdk-17.0.3/lib/server/libjvm.so
#1  0x00007ffff6b6d1ed in LoadedClassesClosure::do_klass(Klass*) ()
   from /scratch/kwalls/jdk/jdk-17.0.3/lib/server/libjvm.so
#2  0x00007ffff66c6016 in ClassLoaderData::loaded_classes_do(KlassClosure*) ()
   from /scratch/kwalls/jdk/jdk-17.0.3/lib/server/libjvm.so
#3  0x00007ffff66c9dd8 in ClassLoaderDataGraph::loaded_classes_do(KlassClosure*) ()
   from /scratch/kwalls/jdk/jdk-17.0.3/lib/server/libjvm.so
#4  0x00007ffff6b6ce6f in JvmtiGetLoadedClasses::getLoadedClasses(JvmtiEnv*, int*, _jclass***) ()
   from /scratch/kwalls/jdk/jdk-17.0.3/lib/server/libjvm.so
#5  0x00007ffff6afa426 in jvmti_GetLoadedClasses () from /scratch/kwalls/jdk/jdk-17.0.3/lib/server/libjvm.so
#6  0x00007ffff7e7561b in classTrack_initialize () from /scratch/kwalls/jdk/jdk-17.0.3/lib/libjdwp.so
#7  0x00007ffff7e778f1 in initialize () from /scratch/kwalls/jdk/jdk-17.0.3/lib/libjdwp.so
#8  0x00007ffff7e783f0 in cbEarlyVMInit () from /scratch/kwalls/jdk/jdk-17.0.3/lib/libjdwp.so
#9  0x00007ffff6b5e756 in JvmtiExport::post_vm_initialized() ()
   from /scratch/kwalls/jdk/jdk-17.0.3/lib/server/libjvm.so
#10 0x00007ffff6f00f54 in Threads::create_vm(JavaVMInitArgs*, bool*) ()
   from /scratch/kwalls/jdk/jdk-17.0.3/lib/server/libjvm.so
#11 0x00007ffff6a007d1 in JNI_CreateJavaVM () from /scratch/kwalls/jdk/jdk-17.0.3/lib/server/libjvm.so
...


LoadedClassesClosure::do_klass does indeed call make_local many times.
If this is the cause of the warning for >400 references, not sure how the code should know how many references to expect (the number of loaded classes..).

If I do make  classTrack_initialize state it wants a large number of local references, that warning disappears:

$ git diff src/jdk.jdwp.agent/share/native/libjdwp/classTrack.c
diff --git a/src/jdk.jdwp.agent/share/native/libjdwp/classTrack.c b/src/jdk.jdwp.agent/share/native/libjdwp/classTrack.c
index e3b30282daa..7dfb9ec01c4 100644
--- a/src/jdk.jdwp.agent/share/native/libjdwp/classTrack.c
+++ b/src/jdk.jdwp.agent/share/native/libjdwp/classTrack.c
@@ -171,6 +171,7 @@ setupEvents()
 void
 classTrack_initialize(JNIEnv *env)
 {
+    WITH_LOCAL_REFS(env, 500) {
     trackingEnv = getSpecialJvmti();
     if (trackingEnv == NULL) {
         EXIT_ERROR(AGENT_ERROR_INTERNAL, "Failed to allocate tag-tracking jvmtiEnv");
@@ -201,4 +202,5 @@ classTrack_initialize(JNIEnv *env)
     } else {
         EXIT_ERROR(error,"loaded classes array");
     }
+    } END_WITH_LOCAL_REFS(env)
 }

**Comment by plummercj** [comment]
This is the problematic code:

```
/* Get all loaded classes (must be inside a WITH_LOCAL_REFS) */
jvmtiError
allLoadedClasses(jclass **ppclasses, jint *pcount)
{
    jvmtiError error;

    *pcount = 0;
    *ppclasses = NULL;
    error = JVMTI_FUNC_PTR(gdata->jvmti,GetLoadedClasses)
                (gdata->jvmti, pcount, ppclasses);
    return error;
}
```

It seems there are a few places where `allLoadedClasses()` is called with either no `WITH_LOCAL_REFS`, or with an inadequate one. However, given that the number of jni localrefs to be allocated is unknown a potentially very large, I'm not sure how callers of JVMTI `GetLoadedClasses()` are suppose to defend against an excess be allocated.

**Review by RealCLanger — COMMENTED** [review]
I think we're getting there. A few further nits and now you also need to update the copyright year.

**Review by sspitsyn — APPROVED** [review]

File: test/jdk/tools/launcher/TestXcheckJNIWarnings.java [hunk 3]
```diff
@@ -25,13 +25,22 @@
...
- * @summary Launching app shouldn't produce any jni warnings.
+ * @summary Launching app with or without jdwp agent shouldn't produce any jni warnings.
  * @modules jdk.compiler
  *          jdk.zipfs
  * @compile TestXcheckJNIWarnings.java
  * @run main TestXcheckJNIWarnings
+ *
+ * @test id=jdwp-agent
+ * @bug 8187442
+ * @summary Launching app with or without jdwp agent shouldn't produce any jni warnings.
```
[the remark this answers is not in the document]
Comment (RealLucy) [reply test/jdk/tools/launcher/TestXcheckJNIWarnings.java:39]: Accepted.

**Comment by kevinjwalls** [comment]
Sounds good.  Comment above showing jvmti_GetLoadedClasses as the cause of the references was just to make sure we knew the cause of that other warning, but I see there is JDK-8296936 to follow up that issue.

File: test/jdk/tools/launcher/TestXcheckJNIWarnings.java [hunk 4]
```diff
@@ -25,14 +25,22 @@
...
  * @run main TestXcheckJNIWarnings
  */
+
+/**
+ * @test id=jdwp-agent
+ * @bug 8187442
+ * @summary Launching app with jdwp agent shouldn't produce any jni warnings.
+ * @modules jdk.compiler
+ *          jdk.zipfs
+ * @run main TestXcheckJNIWarnings -agentlib:jdwp=transport=dt_socket,server=y,suspend=n
```
Comment (dholmes-ora) [inline test/jdk/tools/launcher/TestXcheckJNIWarnings.java:42]: This needed to be `main/othervm`. Not even sure how this can work unless in othervm mode!

Comment (dholmes-ora) [reply test/jdk/tools/launcher/TestXcheckJNIWarnings.java:42]: Sorry ignore that - I see now.

**Pull request closed without merging**